File: ios/Classes/Mappers/BarcodeFormatMapper.swift
import Vision

struct BarcodeFormatMapper {
    static func toVNBarcodeSymbologies(from formats: [BarcodeFormat]) -> [VNBarcodeSymbology] {
        var result: [VNBarcodeSymbology] = []
        
        // Kiểm tra xem có BarcodeFormat.all không
        if formats.contains(.all) {
            // Thêm tất cả các VNBarcodeSymbology nếu .all được chọn
            result.append(contentsOf: [
                .code39, .code39Checksum, .code39FullASCII, .code39FullASCIIChecksum,
                .code93, .code93i, .code128, .i2of5, .i2of5Checksum, .itf14,
                .upce, .ean8, .ean13, .qr, .pdf417, .dataMatrix, .aztec
            ])
            if #available(iOS 15.0, *) {
                result.append(contentsOf: [
                    VNBarcodeSymbology.codabar, VNBarcodeSymbology.gs1DataBar, VNBarcodeSymbology.gs1DataBarExpanded
                ])
            }
        } else {
            // Duyệt qua từng format và thêm VNBarcodeSymbology tương ứng
            for format in formats {
                switch format {
                case .code39:
                    result.append(contentsOf: [
                        .code39, .code39Checksum, .code39FullASCII, .code39FullASCIIChecksum
                    ])
                case .code93:
                    result.append(contentsOf: [.code93, .code93i])
                case .code128:
                    result.append(.code128)
                case .itf:
                    result.append(contentsOf: [.i2of5, .i2of5Checksum, .itf14])
                case .upce:
                    result.append(.upce)
                case .ean8:
                    result.append(.ean8)
                case .ean13:
                    result.append(.ean13)
                case .qr:
                    result.append(.qr)
                case .pdf417:
                    result.append(.pdf417)
                case .dataMatrix:
                    result.append(.dataMatrix)
                case .aztec:
                    result.append(.aztec)
                default:
                    if #available(iOS 15.0, *) {
                        switch format {
                        case BarcodeFormat.codaBar:
                            result.append(VNBarcodeSymbology.codabar)
                            break
                        case BarcodeFormat.gs1DataBar:
                            result.append(VNBarcodeSymbology.gs1DataBar)
                            break
                        case BarcodeFormat.gs1DataBarExtended:
                            result.append(VNBarcodeSymbology.gs1DataBarExpanded)
                            break
                        default:
                            break
                        }
                    }
                    break
                }
            }
        }
        
        return result
    }
    
    static func toBarcodeFormat(from symbology: VNBarcodeSymbology) -> BarcodeFormat {
        switch symbology {
        case .code39, .code39Checksum, .code39FullASCII, .code39FullASCIIChecksum:
            return BarcodeFormat.code39
        case .code93, .code93i:
            return BarcodeFormat.code93
        case .code128:
            return BarcodeFormat.code128
        case .i2of5, .i2of5Checksum, .itf14:
            return BarcodeFormat.itf
        case .upce:
            return BarcodeFormat.upce
        case .ean8:
            return BarcodeFormat.ean8
        case .ean13:
            return BarcodeFormat.ean13
        case .qr:
            return BarcodeFormat.qr
        case .pdf417:
            return BarcodeFormat.pdf417
        case .dataMatrix:
            return BarcodeFormat.dataMatrix
        case .aztec:
            return BarcodeFormat.aztec
        default:
            if #available(iOS 15.0, *) {
                switch symbology {
                case VNBarcodeSymbology.codabar:
                    return .codaBar
                case VNBarcodeSymbology.gs1DataBar:
                    return .gs1DataBar
                case VNBarcodeSymbology.gs1DataBarExpanded:
                    return .gs1DataBarExtended
                default:
                    break
                }
            }
            return .unknown
        }
    }
}

File: android/src/main/kotlin/com/anhlp/barcode_scanner/mappers/BarcodeFormatMapper.kt
package com.anhlp.barcode_scanner.mappers

import com.anhlp.barcode_scanner.Protos

object BarcodeFormatMapper {
    fun toMlKitBarcodeFormats(formats: List<Protos.BarcodeFormat>): List<Int> {
        val results: MutableList<Int> = mutableListOf<Int>()
        if (formats.contains(Protos.BarcodeFormat.all)) {
            results.add(com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ALL_FORMATS)
        } else {
            formats.forEach { format ->
                results.add(toMlKitBarcodeFormat(format))
            }
        }
        return results
    }

    fun toBarcodeFormat(format: Int): Protos.BarcodeFormat {
        return when (format) {
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_39 -> Protos.BarcodeFormat.code39
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_93 -> Protos.BarcodeFormat.code93
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_128 -> Protos.BarcodeFormat.code128
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ITF -> Protos.BarcodeFormat.itf
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_E -> Protos.BarcodeFormat.upce
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_8 -> Protos.BarcodeFormat.ean8
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_13 -> Protos.BarcodeFormat.ean13
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_QR_CODE -> Protos.BarcodeFormat.qr
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_PDF417 -> Protos.BarcodeFormat.pdf417
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_DATA_MATRIX -> Protos.BarcodeFormat.dataMatrix
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_AZTEC -> Protos.BarcodeFormat.aztec
            com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODABAR -> Protos.BarcodeFormat.codaBar
            else -> Protos.BarcodeFormat.unknown
        }
    }

    fun toMlKitBarcodeFormat(format: Protos.BarcodeFormat): Int {
        return when (format) {
            Protos.BarcodeFormat.code39 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_39
            Protos.BarcodeFormat.code93 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_93
            Protos.BarcodeFormat.code128 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_128
            Protos.BarcodeFormat.itf -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ITF
            Protos.BarcodeFormat.upce -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_E
            Protos.BarcodeFormat.ean8 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_8
            Protos.BarcodeFormat.ean13 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_13
            Protos.BarcodeFormat.qr -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_QR_CODE
            Protos.BarcodeFormat.pdf417 -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_PDF417
            Protos.BarcodeFormat.dataMatrix -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_DATA_MATRIX
            Protos.BarcodeFormat.aztec -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_AZTEC
            Protos.BarcodeFormat.codaBar -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODABAR
            else -> com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UNKNOWN
        }
    }


    fun toZXingBarcodeFormats(formats: List<Protos.BarcodeFormat>): List<com.google.zxing.BarcodeFormat> {
        val results: MutableList<com.google.zxing.BarcodeFormat> = mutableListOf()
        formats.forEach { format ->
            val barcodeFormat = toZXingBarcodeFormat(format)
            if (barcodeFormat != null) {
                results.add(barcodeFormat)
            }
        }
        return results
    }

    fun toBarcodeFormat(format: com.google.zxing.BarcodeFormat): Protos.BarcodeFormat {
        return when (format) {
            com.google.zxing.BarcodeFormat.CODE_39 -> Protos.BarcodeFormat.code39
            com.google.zxing.BarcodeFormat.CODE_93 -> Protos.BarcodeFormat.code93
            com.google.zxing.BarcodeFormat.CODE_128 -> Protos.BarcodeFormat.code128
            com.google.zxing.BarcodeFormat.ITF -> Protos.BarcodeFormat.itf
            com.google.zxing.BarcodeFormat.UPC_E -> Protos.BarcodeFormat.upce
            com.google.zxing.BarcodeFormat.EAN_8 -> Protos.BarcodeFormat.ean8
            com.google.zxing.BarcodeFormat.EAN_13 -> Protos.BarcodeFormat.ean13
            com.google.zxing.BarcodeFormat.QR_CODE -> Protos.BarcodeFormat.qr
            com.google.zxing.BarcodeFormat.PDF_417 -> Protos.BarcodeFormat.pdf417
            com.google.zxing.BarcodeFormat.DATA_MATRIX -> Protos.BarcodeFormat.dataMatrix
            com.google.zxing.BarcodeFormat.AZTEC -> Protos.BarcodeFormat.aztec
            com.google.zxing.BarcodeFormat.CODABAR -> Protos.BarcodeFormat.codaBar
            com.google.zxing.BarcodeFormat.RSS_14 -> Protos.BarcodeFormat.gs1DataBar
            com.google.zxing.BarcodeFormat.RSS_EXPANDED -> Protos.BarcodeFormat.gs1DataBarExtended
            else -> Protos.BarcodeFormat.unknown
        }
    }

    fun toZXingBarcodeFormat(format: Protos.BarcodeFormat): com.google.zxing.BarcodeFormat? {
        return when (format) {
            Protos.BarcodeFormat.code39 -> com.google.zxing.BarcodeFormat.CODE_39
            Protos.BarcodeFormat.code93 -> com.google.zxing.BarcodeFormat.CODE_93
            Protos.BarcodeFormat.code128 -> com.google.zxing.BarcodeFormat.CODE_128
            Protos.BarcodeFormat.itf -> com.google.zxing.BarcodeFormat.ITF
            Protos.BarcodeFormat.upce -> com.google.zxing.BarcodeFormat.UPC_E
            Protos.BarcodeFormat.ean8 -> com.google.zxing.BarcodeFormat.EAN_8
            Protos.BarcodeFormat.ean13 -> com.google.zxing.BarcodeFormat.EAN_13
            Protos.BarcodeFormat.qr -> com.google.zxing.BarcodeFormat.QR_CODE
            Protos.BarcodeFormat.pdf417 -> com.google.zxing.BarcodeFormat.PDF_417
            Protos.BarcodeFormat.dataMatrix -> com.google.zxing.BarcodeFormat.DATA_MATRIX
            Protos.BarcodeFormat.aztec -> com.google.zxing.BarcodeFormat.AZTEC
            Protos.BarcodeFormat.codaBar -> com.google.zxing.BarcodeFormat.CODABAR
            Protos.BarcodeFormat.gs1DataBar -> com.google.zxing.BarcodeFormat.RSS_14
            Protos.BarcodeFormat.gs1DataBarExtended -> com.google.zxing.BarcodeFormat.RSS_EXPANDED
            else -> null
        }
    }
}
